src/main/java/uk/gov/hmcts/darts/courthouse/service/CourthouseServiceImpl.java

```diff
@@ -121,6 +135,150 @@ public CourthouseEntity addCourtHouse(Courthouse courthouse) {
+        }
+
+        //ensure if security roles exist, TRANSCRIBER role also exists
+        if (securityRoleTranscriber != null && (!securityRoleIds.isEmpty() && !securityRoleIds.contains(securityRoleTranscriber.getId()))) {
+                throw new DartsApiException(AuthorisationError.BAD_REQUEST_ONLY_TRANSCIBER_ROLES_MAY_BE_ASSIGNED);
+
+        }
+
+        CourthouseEntity mappedEntity = getMappedEntityWithRegion(courthousePost);
+
+        Set<SecurityGroupEntity> securityGroupEntities = getSecurityGroupEntities(courthousePost, mappedEntity,
+                                                                                  securityRoleIds, securityRoleRequester,
+                                                                                  securityRoleApprover);
+        mappedEntity.setSecurityGroups(securityGroupEntities);
+
+        courthouseRepository.saveAndFlush(mappedEntity);
+
+        return getExtendedCourthousePost(courthousePost, mappedEntity);
+    }
+
+    private ExtendedCourthousePost getExtendedCourthousePost(CourthousePost courthousePost, CourthouseEntity mappedEntity) {
+        ExtendedCourthousePost extendedCourthousePost = courthousePostMapper.mapFromEntityToExtendedCourthousePost(courthousePost);
+
+        extendedCourthousePost.setId(mappedEntity.getId());
+        extendedCourthousePost.setCreatedDateTime(mappedEntity.getCreatedDateTime());
+        extendedCourthousePost.setLastModifiedDateTime(mappedEntity.getLastModifiedDateTime());
+        extendedCourthousePost.securityGroupIds(securityGroupIdMapper.mapSecurityGroupEntitiesToIds(mappedEntity.getSecurityGroups()));
+        extendedCourthousePost.setRegionId(mappedEntity.getRegion() == null ? null : mappedEntity.getRegion().getId());
+
+        return extendedCourthousePost;
+    }
+
+    private CourthouseEntity getMappedEntityWithRegion(CourthousePost courthousePost) {
+        CourthouseEntity mappedEntity = courthousePostMapper.mapToEntity(courthousePost);
+
+        //ensure region exists in dB
+        List<RegionEntity> regions = regionRepository.findAll();
+        if (CollectionUtils.isNotEmpty(regions) && courthousePost.getRegionId() != null) {
+            List<Integer> regionIds = regions.stream().map(RegionEntity::getId).toList();
+
+            if (CollectionUtils.isNotEmpty(regionIds) && !regionIds.contains(courthousePost.getRegionId())) {
+                throw new DartsApiException(AuthorisationError.BAD_REQUEST_REGION_ID_DOES_NOT_EXIST);
+            }
+
+            //get region
+            RegionEntity regionFound = null;
+            for (RegionEntity region : regions) {
+                if (region.getId().equals(courthousePost.getRegionId())) {
+                    regionFound = region;
+                }
+            }
+
+            mappedEntity.setRegion(regionFound);
+        }
+        return mappedEntity;
```

Comment:
I think this can be simplified considerably by calling the `RegionRepository.findById()` method, rather that getting them all and doing your own linear search.

Reply:
RegionRepository.findById() returns Optional. This led to several issues that would need more time to fix than I had today.

Reply:
I have addressed this